File: apps/macos/LexiconBar/Sources/LexiconBar/FixEverywhere/NormalizeClient.swift
import Foundation
import LexiconBarKit

/// Talks to `lexicon serve` (`POST /normalize`, `GET /health`) with the
/// bearer token from `~/.config/lexicon/serve.json`. The token file is read
/// lazily and re-read after any failure, so a `lexicon serve --uninstall &&
/// --install` (new token) heals on the next burst.
final class NormalizeClient: @unchecked Sendable {
    struct Credentials: Equatable {
        let port: Int
        let token: String
        var baseURL: URL { URL(string: "http://127.0.0.1:\(port)")! }
    }

    enum Failure: Error, CustomStringConvertible {
        case noCredentials(String)
        case transport(String)
        case http(Int, String)
        case badBody

        var description: String {
            switch self {
            case .noCredentials(let path): return "serve.json not found at \(path); run `lexicon serve --install`."
            case .transport(let why): return "Local API unreachable: \(why)"
            case .http(let code, let body): return "Local API returned \(code)\(body.isEmpty ? "" : ": \(body)")"
            case .badBody: return "Local API returned an unexpected body."
            }
        }
    }

    let timeout: TimeInterval
    private let session: URLSession
    private let lock = NSLock()
    private var cached: Credentials?

    init(timeout: TimeInterval = 1.5) {
        self.timeout = timeout
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = timeout
        config.timeoutIntervalForResource = timeout + 0.5
        config.waitsForConnectivity = false
        session = URLSession(configuration: config)
    }

    /// `$LEXICON_PATH`'s directory, else `$XDG_CONFIG_HOME/lexicon`, else `~/.config/lexicon`.
    static var serveJSONPath: String {
        let env = ProcessInfo.processInfo.environment
        if let lexiconPath = env["LEXICON_PATH"], !lexiconPath.isEmpty {
            return (lexiconPath as NSString).deletingLastPathComponent + "/serve.json"
        }
        if let xdg = env["XDG_CONFIG_HOME"], !xdg.isEmpty {
            return xdg + "/lexicon/serve.json"
        }
        return FileManager.default.homeDirectoryForCurrentUser.path + "/.config/lexicon/serve.json"
    }

    static func readCredentials(path: String = serveJSONPath) -> Credentials? {
        guard let data = FileManager.default.contents(atPath: path),
              let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
              let token = object["token"] as? String, !token.isEmpty else { return nil }
        let port = (object["port"] as? NSNumber)?.intValue ?? 41733
        return Credentials(port: port, token: token)
    }

    func credentials(reload: Bool = false) -> Credentials? {
        lock.lock(); defer { lock.unlock() }
        if !reload, let cached { return cached }
        cached = NormalizeClient.readCredentials()
        return cached
    }

    func invalidateCredentials() {
        lock.lock(); cached = nil; lock.unlock()
    }

    /// POST /normalize with `text`. Completion runs on an arbitrary queue.
    func normalize(_ text: String, completion: @escaping (Result<NormalizeResponse, Failure>) -> Void) {
        guard let creds = credentials() else {
            completion(.failure(.noCredentials(NormalizeClient.serveJSONPath)))
            return
        }
        var request = URLRequest(url: creds.baseURL.appendingPathComponent("normalize"))
        request.httpMethod = "POST"
        request.timeoutInterval = timeout
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(creds.token)", forHTTPHeaderField: "Authorization")
        request.httpBody = try? JSONSerialization.data(withJSONObject: ["text": text])
        session.dataTask(with: request) { [weak self] data, response, error in
            if let error {
                self?.invalidateCredentials()
                completion(.failure(.transport(error.localizedDescription)))
                return
            }
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            guard status == 200 else {
                if status == 401 { self?.invalidateCredentials() }
                let body = data.map { String(decoding: $0.prefix(200), as: UTF8.self) } ?? ""
                completion(.failure(.http(status, body)))
                return
            }
            guard let data, let parsed = NormalizeResponse.parse(data) else {
                completion(.failure(.badBody))
                return
            }
            completion(.success(parsed))
        }.resume()
    }

    /// GET /health, synchronously (for `--status`). Returns the term count.
    func healthSync(timeout: TimeInterval = 2) -> Result<Int, Failure> {
        let port = credentials()?.port ?? 41733
        var request = URLRequest(url: URL(string: "http://127.0.0.1:\(port)/health")!)
        request.timeoutInterval = timeout
        let done = DispatchSemaphore(value: 0)
        var outcome: Result<Int, Failure> = .failure(.transport("no response"))
        session.dataTask(with: request) { data, response, error in
            defer { done.signal() }
            if let error { outcome = .failure(.transport(error.localizedDescription)); return }
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            guard status == 200, let data,
                  let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
                outcome = .failure(.http(status, ""))
                return
            }
            outcome = .success((object["terms"] as? NSNumber)?.intValue ?? 0)
        }.resume()
        _ = done.wait(timeout: .now() + timeout + 1)
        return outcome
    }
}

File: apps/windows/src/LexiconBar.App/NormalizeClient.cs
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

namespace LexiconBar;

/// <summary>
/// Talks to <c>lexicon serve</c> on loopback, with the bearer token from
/// <c>serve.json</c>. The token file is read lazily and re-read after any
/// failure, so restarting the server with a fresh token heals on the next burst
/// instead of needing the tray app restarted.
///
/// A direct port of <c>NormalizeClient.swift</c> apart from where the file
/// lives; see <see cref="ServePaths"/> for that, because the answer is not the
/// one you would guess.
/// </summary>
public sealed class NormalizeClient : IDisposable
{
    public abstract record Failure(string Description)
    {
        public sealed record NoCredentials(string SearchedPath) : Failure(
            $"serve.json not found at {SearchedPath}; run `lexicon serve` on this machine.");

        public sealed record Transport(string Why) : Failure($"Local API unreachable: {Why}");

        public sealed record Http(int Status, string Body) : Failure(
            $"Local API returned {Status}{(Body.Length == 0 ? "" : ": " + Body)}");

        public sealed record BadBody() : Failure("Local API returned an unexpected body.");
    }

    public sealed record Result(NormalizeResponse? Response, Failure? Failure)
    {
        public bool Ok => Response is not null;
    }

    private readonly HttpClient _http;
    private readonly object _gate = new();
    private ServeCredentials? _cached;
    private string? _cachedPath;

    public NormalizeClient(TimeSpan? timeout = null)
    {
        _http = new HttpClient(new SocketsHttpHandler
        {
            // Loopback only, no proxy, no connection reuse surprises.
            UseProxy = false,
            AllowAutoRedirect = false,
            ConnectTimeout = TimeSpan.FromMilliseconds(500),
        })
        {
            Timeout = timeout ?? TimeSpan.FromSeconds(1.5),
        };
    }

    /// <summary>The path <c>serve.json</c> was last found at, for the doctor window.</summary>
    public string? ResolvedPath => _cachedPath;

    public ServeCredentials? Credentials(bool reload = false)
    {
        lock (_gate)
        {
            if (!reload && _cached is not null) return _cached;
            _cached = null;
            _cachedPath = null;

            IReadOnlyDictionary<string, string?> env = ServePaths.CurrentEnvironment();
            string home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
            foreach (string candidate in ServePaths.Candidates(env, home))
            {
                try
                {
                    if (!File.Exists(candidate)) continue;
                    if (ServeCredentials.Parse(File.ReadAllText(candidate)) is not ServeCredentials parsed) continue;
                    _cached = parsed;
                    _cachedPath = candidate;
                    return parsed;
                }
                catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
                {
                    // Being read by the server right now; try the next one and
                    // then try again on the next burst.
                }
            }

            return null;
        }
    }

    public void InvalidateCredentials()
    {
        lock (_gate)
        {
            _cached = null;
        }
    }

    /// <summary><c>POST /normalize</c>.</summary>
    public async Task<Result> NormalizeAsync(string text, CancellationToken cancellation = default)
    {
        if (Credentials() is not ServeCredentials credentials)
        {
            IReadOnlyDictionary<string, string?> env = ServePaths.CurrentEnvironment();
            string home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
            return new Result(null, new Failure.NoCredentials(ServePaths.Candidates(env, home).FirstOrDefault() ?? "(none)"));
        }

        using HttpRequestMessage request = new(HttpMethod.Post, new Uri(credentials.BaseUri, "normalize"))
        {
            Content = new StringContent(
                JsonSerializer.Serialize(new Dictionary<string, string> { ["text"] = text }),
                Encoding.UTF8,
                "application/json"),
        };
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", credentials.Token);

        HttpResponseMessage response;
        try
        {
            response = await _http.SendAsync(request, cancellation).ConfigureAwait(false);
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException or OperationCanceledException)
        {
            InvalidateCredentials();
            return new Result(null, new Failure.Transport(ex.Message));
        }

        using (response)
        {
            if (response.StatusCode != HttpStatusCode.OK)
            {
                if (response.StatusCode == HttpStatusCode.Unauthorized) InvalidateCredentials();
                string body = await ReadCappedAsync(response, cancellation).ConfigureAwait(false);
                return new Result(null, new Failure.Http((int)response.StatusCode, body));
            }

            string json = await response.Content.ReadAsStringAsync(cancellation).ConfigureAwait(false);
            return NormalizeResponse.Parse(json) is NormalizeResponse parsed
                ? new Result(parsed, null)
                : new Result(null, new Failure.BadBody());
        }
    }

    /// <summary><c>GET /health</c>. Returns the term count, for the doctor window.</summary>
    public async Task<int?> HealthAsync(CancellationToken cancellation = default)
    {
        if (Credentials() is not ServeCredentials credentials) return null;
        try
        {
            using HttpRequestMessage request = new(HttpMethod.Get, new Uri(credentials.BaseUri, "health"));
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", credentials.Token);
            using HttpResponseMessage response = await _http.SendAsync(request, cancellation).ConfigureAwait(false);
            if (response.StatusCode != HttpStatusCode.OK) return null;

            string json = await response.Content.ReadAsStringAsync(cancellation).ConfigureAwait(false);
            using JsonDocument document = JsonDocument.Parse(json);
            return document.RootElement.TryGetProperty("terms", out JsonElement terms)
                   && terms.TryGetInt32(out int count)
                ? count
                : 0;
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException
                                      or OperationCanceledException or JsonException)
        {
            return null;
        }
    }

    /// <summary>
    /// <c>POST /add {canonical, never:[original]}</c> — the bubble's Never button.
    /// </summary>
    public Task<bool> NeverAsync(string canonical, string original, CancellationToken cancellation = default) =>
        PostAsync("add", new Dictionary<string, object>
        {
            ["canonical"] = canonical,
            ["never"] = new[] { original },
        }, cancellation);

    /// <summary>
    /// <c>POST /learn {heard, meant}</c> — the bubble's Add button.
    /// </summary>
    public Task<bool> LearnAsync(string heard, string meant, CancellationToken cancellation = default) =>
        PostAsync("learn", new Dictionary<string, object>
        {
            ["heard"] = heard,
            ["meant"] = meant,
        }, cancellation);

    private async Task<bool> PostAsync(
        string route,
        Dictionary<string, object> body,
        CancellationToken cancellation)
    {
        if (Credentials() is not ServeCredentials credentials) return false;
        try
        {
            using HttpRequestMessage request = new(HttpMethod.Post, new Uri(credentials.BaseUri, route))
            {
                Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json"),
            };
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", credentials.Token);
            using HttpResponseMessage response = await _http.SendAsync(request, cancellation).ConfigureAwait(false);
            return response.IsSuccessStatusCode;
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException or OperationCanceledException)
        {
            InvalidateCredentials();
            return false;
        }
    }

    private static async Task<string> ReadCappedAsync(HttpResponseMessage response, CancellationToken cancellation)
    {
        try
        {
            string body = await response.Content.ReadAsStringAsync(cancellation).ConfigureAwait(false);
            return body.Length <= 200 ? body : body[..200];
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException or OperationCanceledException)
        {
            return string.Empty;
        }
    }

    public void Dispose() => _http.Dispose();
}
